panoptes_aggregation/tests/batch_aggregation/test_batch_aggregation.py

```diff
@@ -0,0 +1,124 @@
+import unittest
+from unittest.mock import patch, MagicMock, call
+from panoptes_aggregation.scripts import batch_utils
+from panoptes_aggregation.batch_aggregation import run_aggregation
+from panoptes_aggregation import batch_aggregation as batch_agg
+
+wf_export = 'panoptes_aggregation/tests/batch_aggregation/wf_export.csv'
+cls_export = 'panoptes_aggregation/tests/batch_aggregation/cls_export.csv'
+
+
+@patch("panoptes_aggregation.batch_aggregation.BatchAggregator._connect_api_client", new=MagicMock())
+class TestBatchAggregation(unittest.TestCase):
+    @patch("panoptes_aggregation.batch_aggregation.workflow_extractor_config")
+    @patch("panoptes_aggregation.batch_aggregation.BatchAggregator")
+    def test_run_aggregation(self, mock_aggregator, mock_wf_ext_conf):
+        mock_aggregator.process_wf_export.return_value = MagicMock()
+        mock_aggregator.process_cls_export.return_value = MagicMock()
+
+        mock_df = MagicMock()
+        test_extracts = {'question_extractor': mock_df}
+        batch_utils.batch_extract = MagicMock(return_value=test_extracts)
+        mock_reducer = MagicMock()
+        batch_utils.batch_reduce = mock_reducer
+
+        run_aggregation(1, 10, 100)
+        mock_aggregator.assert_called_once_with(1, 10, 100)
+        mock_wf_ext_conf.assert_called_once()
+        batch_utils.batch_extract.assert_called_once()
+        mock_df.to_csv.assert_called()
+        batch_utils.batch_reduce.assert_called()
+        self.assertEqual(mock_reducer.call_count, 2)
+
+        # The reducer's call list includes subsequent calls to to_csv, but the args are methods called on the mock
+        # rather than use the set values i.e. "<MagicMock name='BatchAggregator().output_path' id='140281634764400'>"
+        # mock_aggregator.workflow_id = '10'
+        # mock_aggregator.output_path = 'tmp/10'
+        # mock_reducer.assert_has_calls([
+        #         call(mock_df, {'reducer_config': {'question_reducer': {}}}),
+        #         call().to_csv('tmp/10/10_reducers.csv', mode='a'),
+        #         call(mock_df, {'reducer_config': {'question_consensus_reducer': {}}}),
+        #         call().to_csv('tmp/10/10_reducers.csv', mode='a'),
+        #     ])
+
+        # How do I test the specific instance of BatchAggregator rather than the mocked class?
+        # mock_aggregator.upload_files.assert_called_once()
```

Comment:
You can write a test that rather than mock the entire class just mocks the method you wish to test
```python
@patch("panoptes_aggregation.batch_aggregation.BatchAggregator.upload_files")
@patch(...)
```

Reply:
I'm already mocking the whole BatchAggregator, but I think my issue is that it's mocking a class method when what I really wanna do is set a mock a method on the specific instance of BatchAggregator that the `run_aggregation` method is instantiating. Mocking `BatchAggregator.upload_files` like above also doesn't work.  I put one PDB breakpoint in the method itself and one in the spec:

```python
> /usr/src/aggregation/panoptes_aggregation/batch_aggregation.py(55)run_aggregation()
-> ba.update_panoptes()
(Pdb) ba
<MagicMock name='BatchAggregator()' id='139923834173280'>
(Pdb) ba.upload_files
<MagicMock name='BatchAggregator().upload_files' id='139923833924432'>
(Pdb) ba.upload_files()
<MagicMock name='BatchAggregator().upload_files()' id='139923833994928'>
(Pdb) continue
>> PDB continue (IO-capturing resumed)
>> PDB set_trace (IO-capturing turned off) 
/usr/src/aggregation/panoptes_aggregation/tests/batch_aggregation/test_batch_aggregation.py(45)test_run_aggregation()
-> mock_uploader.assert_called_once()
(Pdb) mock_uploader
<MagicMock name='upload_files' id='139923834625088'>
(Pdb) mock_uploader()
<MagicMock name='upload_files()' id='139923834173328'>
```
Maybe because the thing I'm mocking isn't actually exactly the thing I'm calling?

Reply:
Figured it out! You can use `mocked_class_instance = mocked_class.return_value` to get access to an "instance" of the class, then set mocks on instance methods. Updated a few tests with this syntax.